AsyncDisplayKit/Private/ASCollectionDataInternal.mm

```diff
@@ -0,0 +1,320 @@
+                       withIdentifier:(ASItemIdentifier)identifier
+                                index:(NSInteger)index
+                            nodeBlock:(ASCellNodeBlock)nodeBlock
+{
+  ASCollectionSectionImpl *section = _currentSection;
+  if (section == nil) {
+    section = [self _sectionWithIdentifier:identifier appendingIfCreated:YES];
+  }
+  ASCollectionItemImpl *item = [self _itemWithIdentifier:identifier nodeBlock:nodeBlock];
+  [_supplementaryElementKinds addObject:elementKind];
+  [section setSupplementaryElement:item ofKind:elementKind atIndex:index];
+  _usedItems[identifier] = item;
+}
+
+#pragma mark - Item / Section Access (Public)
+
+- (id<ASCollectionItem>)itemWithIdentifier:(ASItemIdentifier)identifier nodeBlock:(nonnull ASCellNodeBlock)nodeBlock
+{
+  ASDisplayNodeAssertNil(_usedItems[identifier], @"Attempt to use the same item twice. Identifier: %@", identifier);
+  ASCollectionItemImpl *item = [self _itemWithIdentifier:identifier nodeBlock:nodeBlock];
+  _usedItems[identifier] = item;
+  return item;
+}
+
+- (id<ASCollectionSection>)sectionWithIdentifier:(ASSectionIdentifier)identifier
+{
+  ASDisplayNodeAssertNil(_usedSections[identifier], @"Attempt to use the same section twice. Identifier: %@", identifier);
+  return [self _sectionWithIdentifier:identifier appendingIfCreated:NO];
+}
+
+#pragma mark - Framework Accessors
+
+- (NSMutableSet *)supplementaryElementKinds
+{
+  return [_supplementaryElementKinds mutableCopy];
+}
+
+- (NSArray *)sectionsInternal
+{
+  return _mutableSections;
+}
+
+- (ASCollectionItemImpl *)elementOfKind:(ASSupplementaryElementKind)kind atIndexPath:(NSIndexPath *)indexPath
+{
+  ASCollectionSectionImpl *section = self.sectionsInternal[indexPath.section];
+  if (kind == ASDataControllerRowNodeKind) {
+    return section.itemsInternal[indexPath.item];
+  } else {
+    return section.supplementaryElements[kind][@(indexPath.item)];
+  }
+}
+
+- (void)markCompleted
+{
+  _completed = YES;
+}
+
+- (std::vector<NSInteger>)itemCounts
+{
+  std::vector<NSInteger> result;
+  for (ASCollectionSectionImpl *section in self.sectionsInternal) {
+    result.push_back(section.itemsInternal.count);
+  }
+  return result;
+}
+
+- (NSString *)description
+{
+  return ASObjectDescriptionMake(self, [self propertiesForDescription]);
+}
+
+- (NSMutableArray <NSDictionary *> *)propertiesForDescription
+{
+  NSMutableArray *array = [NSMutableArray array];
+  [array addObject:@{ @"sections" : _mutableSections }];
+  return array;
+}
+
+#pragma mark - Private
+
+- (ASCollectionItemImpl *)_itemWithIdentifier:(ASItemIdentifier)identifier nodeBlock:(nonnull ASCellNodeBlock)nodeBlock
+{
+  ASCollectionItemImpl *item = _itemsDict[identifier];
+  if (item == nil) {
+    void (^postNodeBlock)(ASCellNode *) = _postNodeBlock;
+    item = [[ASCollectionItemImpl alloc] initWithIdentifier:identifier nodeBlock:^{
+      ASCellNode *node = nodeBlock();
+      if (postNodeBlock != nil) {
+        postNodeBlock(node);
```

Comment:
is postNodeBlock always / only internal?  The name seems a bit ambiguous -- is it postNodeAllocationBlock?  nodeSetupBlock?